Port UriTypeConverter source and add tests.

ianhays opened the pull request:
The first commit ports the sources from TFS by dotnet-bot.

The second commit clean ups and changes the sources to work in CoreFX. It also adds some tests. I couldn't find the existing tests so I just wrote some new ones in the style of the other tests in the project. The unusual comment style of the sources (devdoc) is to match the other sources in the TypeConverter namespace for consistency.

resolves #7501

@Tanya-Solyanik @stephentoub

justinvp commented on src/System.ComponentModel.TypeConverter/src/System/ComponentModel/UriTypeConverter.cs:
```diff
@@ -0,0 +1,92 @@
+            if (sourceType == typeof(string) || sourceType == typeof(Uri))
+                return true;
+
+            return false;
```
Nit: This could be simplified as:

```csharp
return sourceType == typeof(string) || sourceType == typeof(Uri);
```

justinvp commented on src/System.ComponentModel.TypeConverter/src/System/ComponentModel/UriTypeConverter.cs:
```diff
@@ -0,0 +1,92 @@
+            if (destinationType == typeof(Uri))
+                return true;
+
+            return false;
```
Nit: The whole method could be simplified as:

```csharp
public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType) =>
    destinationType == typeof(string) || destinationType == typeof(Uri);
```

justinvp commented on src/System.ComponentModel.TypeConverter/src/System/ComponentModel/UriTypeConverter.cs:
```diff
@@ -0,0 +1,92 @@
+            string uriString = value as string;
+            if (uriString != null)
+            {
+                if (String.IsNullOrEmpty(uriString))
```
Nit: lowercase `string`, e.g.:

```csharp
if (string.IsNullOrEmpty(uriString))
```

justinvp commented on src/System.ComponentModel.TypeConverter/src/System/ComponentModel/UriTypeConverter.cs:
```diff
@@ -0,0 +1,92 @@
+                return new Uri(uriString, UriKind.RelativeOrAbsolute);
+            }
+
+            Uri valueAsUri = value as Uri;
```
Nit: Should this be named `uri` to be consistent the naming used below in `ConvertTo`?

justinvp commented on src/System.ComponentModel.TypeConverter/src/System/ComponentModel/UriTypeConverter.cs:
```diff
@@ -0,0 +1,92 @@
+
+            Uri uri = value as Uri;
+
+            if (uri != null)
```
Nit: In `ConvertFrom` above, there is no extra line between the `Uri uri = value as Uri;` and the `if (uri != null)`. Ideally both methods would be consistent.

hughbe commented:
Does the `ref` folder need to be updated?

justinvp commented:
#7501 mentions the following snippet:

```csharp
TypeDescriptor.GetConverter(typeof(Uri)).ConvertFromInvariantString("http://www.bing.com");
```

Does `TypeDescriptor.GetConverter` need to be wired up to support `UriTypeConverter`?

Looks like it'd need to be added [here](https://github.com/dotnet/corefx/blob/6eed3985aa246946ff66016a4972d55525098f9e/src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L48).

ianhays commented:
>Does TypeDescriptor.GetConverter need to be wired up to support UriTypeConverter? Or will it just work?

Yes, it does. Thanks, I forgot that bit. I'll add it to the tests for that as well.

>Does the ref folder need to be updated? (I don't fully understand how this folder fits in here)

For some reason I assumed that was done automatically now but I admittedly haven't added API in a while so I'm a bit out of the loop on that. @mellinoe do we still have to modify the ref file to add any new surface area?

ianhays commented:
test Innerloop OSX Debug Build and Test please

ianhays commented:
Update: we did have to add uritypeconverter to the ref file for the net45 build. I updated this and rebased to include that change. It also required I keep UriTypeConverter in the ```System``` namespace (rather than System.ComponentModel) so I changed that too.

ianhays commented:
test Innerloop OSX Debug Build and Test please
test Innerloop OSX Release Build and Test please

ianhays commented:
test Innerloop OSX Debug Build and Test please
test Innerloop OSX Release Build and Test please

stephentoub commented:
LGTM

stephentoub commented:
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please
